packages/flutter/test/widgets/app_lifecycle_listener_test.dart

```diff
@@ -0,0 +1,155 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:ui';
+
+import 'package:flutter/services.dart';
+import 'package:flutter/widgets.dart';
+import 'package:flutter_test/flutter_test.dart';
+
+void main() {
+  Future<void> setAppLifeCycleState(AppLifecycleState state) async {
+    final ByteData? message = const StringCodec().encodeMessage(state.toString());
+    await ServicesBinding.instance.defaultBinaryMessenger.handlePlatformMessage('flutter/lifecycle', message, (_) {});
+  }
+
+  Future<void> sendAppExitRequest() async {
+    final ByteData message = const JSONMethodCodec().encodeMethodCall(const MethodCall('System.requestAppExit'));
+    await ServicesBinding.instance.defaultBinaryMessenger.handlePlatformMessage('flutter/platform', message, (_) {});
+  }
+
+  setUp(() async {
+    WidgetsFlutterBinding.ensureInitialized();
+    WidgetsBinding.instance
+      ..resetEpoch()
+      ..platformDispatcher.onBeginFrame = null
+      ..platformDispatcher.onDrawFrame = null;
+    final TestWidgetsFlutterBinding binding = TestWidgetsFlutterBinding.instance;
+    binding.platformDispatcher.resetInitialLifecycleState();
+    binding.readTestInitialLifecycleStateFromNativeWindow();
+    // Reset the state to detached. Going to paused first makes it a valid
+    // transition from any state.
+    await setAppLifeCycleState(AppLifecycleState.paused);
+    await setAppLifeCycleState(AppLifecycleState.detached);
+  });
+
+  testWidgets('listens to AppLifecycleState', (WidgetTester tester) async {
+    AppLifecycleState lastState = tester.binding.lifecycleState!;
+    void stateChange(AppLifecycleState state) {
+      lastState = state;
```

Comment:
should we maybe store them in a list to verify that it doesn't accidentally trigger more then once for a transition?

Reply:
Also (unless it is coming later): Should we test some transitions that trigger multiple state changes from the `_generateStateTransitions` code in the binding?

Reply:
Changing to a list will already verify the multiple changes, since `detached` -> `inactive` will generate a `resumed` in between. Done.